entities/PropertyManagementService.js

```diff
@@ -0,0 +1,199 @@
+  }
+
+  static demolish(gameState, property) {
+    gameState.config.propertyConfig[
+      property.buildings !== 5 ? 'houses' : 'hotels'
+    ] += 1;
+    property.buildings -= 1;
+    WealthService.sellAsset(gameState.currentPlayer, property.houseCost / 2);
+  }
+
+  static toggleMortgageOnProperty(
+    gameState,
+    boardProperty,
+    player = gameState.currentPlayer
+  ) {
+    boardProperty.mortgaged = !boardProperty.mortgaged;
+    const INTEREST_RATE = 0.1;
+    const mortgageBaseCost = boardProperty.price / 2;
+
+    if (boardProperty.mortgaged) {
+      WealthService.sellAsset(player, mortgageBaseCost);
+    } else {
+      WealthService.decrement(player, mortgageBaseCost * INTEREST_RATE);
+      WealthService.buyAsset(player, mortgageBaseCost);
+    }
+  }
+
+  static hasMonopoly(gameState, propertyGroup, playerId) {
+    return this.getProperties(gameState)
+      .filter((p) => p.group === propertyGroup)
+      .every((p) => p.ownedBy === playerId);
+  }
+
+  static getAvailableManagementActions(gameState) {
+    let availableActions = ['TOGGLE_MORTGAGE'];
+
+    if (this.getRenoProperties(gameState).length > 0) {
+      availableActions.push('RENOVATE');
+    }
+
+    if (this.getDemoProperties(gameState).length > 0) {
+      availableActions.push('DEMOLISH');
+    }
+
+    availableActions.push('CANCEL');
+    return availableActions;
+  }
+
+  static getRenoProperties(gameState) {
+    const { houses, hotels, properties } = gameState.config.propertyConfig;
+    const player = gameState.currentPlayer;
+    if (houses + hotels === 0) return [];
+    const limitedHouses = houses === 0;
+    const limitedHotels = hotels === 0;
+
+    const availableRenoProps = _(properties)
+      .filter((p) => p.ownedBy === player.id)
+      .groupBy((p) => p.group) // { 'groupName': [properties] }
+      .mapValues((properties) => ({
+        properties,
+        hasMonopoly: this.hasMonopoly(
+          gameState,
+          properties[0].group,
+          properties[0].ownedBy
+        ),
+        minBuildingCount: Math.min.apply(
+          Math,
+          properties.map((p) => p.buildings)
+        ),
+      })) // { 'groupName': { propertiesInTheGroup: [], hasMonopoly: bool, minBuildingCount: number } }
+      // CHECK: Property must be part of a monopoly
+      .filter((pg) => pg.hasMonopoly)
+      // CHECK: Minimum building count
+      .mapValues((modifiedProperties) => ({
+        properties: modifiedProperties.properties.filter(
+          (p) => p.buildings === modifiedProperties.minBuildingCount
+        ),
+      }))
+      .map((pg) => pg.properties) // [ [ { propertyGroup1 }], [ { propertyGroup2} ] ]
+      .flattenDeep()
+      .filter((p) => !p.mortgaged)
+      // CHECK: Must have houses to build
+      .filter((p) => !limitedHouses || (limitedHouses && p.buildings === 4))
+      // CHECK: Must have hotels to build
+      .filter((p) => !limitedHotels || (limitedHotels && p.buildings < 4))
```

Comment:
Why do both of these checks happen in this flow ? Shouldn't they be contextual ?

Reply:
How would they be contextual? The limitedHouses/Hotels are a global quantity.

Reply:
`limitedHouses ? p.buildings === 4 : p.buildings < 4` ?

Reply:
`limitedHouses || p.buildings === 4` ?